### ISSUE TITLE
Support distributed prefill with local decode

### ISSUE THREAD

**Issue by lobanov** [antirez/ds4#304]
It should be possible to run generation locally on one device even if prefill is distributed. For example, it may be practical in a setup that uses DGX Spark to offload some of even all the prefill and then a Mac for generation. DGX Spark has a very fast GPU in theory, and Mac has very fast memory, so offloading prefill to the former while keeping all the generation on the latter could be the way to speed up large-context generation. This is also practical on homogeneous hardware, such as two Macs, because local decode is faster.

Conceptually it's similar to [Kimi's Mooncake architecture](https://arxiv.org/abs/2407.00079), but at much smaller scale.

**Goal**: allow ds4 to pipeline prefill across two nodes while having the node owning the `output` layer to decode fully locally.

**Comment by lobanov** [comment]
Some back of the napkin calculation for feasibility:

Pipelining of prefills is already an existing mechanism, the only additional overhead is shipping KV cache for earlier (or all layers) to the machine doing the inference. It makes sense to do that per chunk alongside the layer activation data being pipelined anyway. If full 1M tokens context is taking 26GB (from README), then assuming context chunking by 4096 tokens, full context is 244 chunks with ~106MB per activation chunk. Practical bandwidth of 2.5GbE connection is around 280Mbytes/s with 1-2ms RTT. Activation chunk itself is around 64MB (4096 tokens × hidden size 4096 × 4 bytes FP32). Sending 170MB (chunk + KV cache) instead of 64MB over that link adds only extra 0.4-0.5s latency. Since inference is strictly feed-forward, and assuming efficient pipelining implementation, that 0.5s latency should be added only once per prefill.

This makes it feasible.

On Thunderbolt 5 vs 2.5Gbps, I'm yet to test this, but I suspect practical difference is negligible here. It's reported (from README) that setting `--dist-activation-bits` to a smaller value doesn't provide significant improvement despite cutting the payload size in half (if set to 16), from which I conclude that distributed inference at this scale is not bandwidth-bound. Will report back after practical testing once some hardware arrives later this week.

**Comment by lobanov** [comment]
I did some basic testing to prove feasibility over 5GbE direct link.

Setup:
Code `main`@`ba00a8a`
Model `DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix.gguf` 
1) DGX Spark as worker node:
`./ds4 --role worker --layers 22:output --coordinator 10.77.0.1 1234`
2) MacBook M5 Max 128GB as coordinator:
`./ds4 --role coordinator --layers 0:21 --listen 10.77.0.1 1234 --prompt-file README.md --nothink -n 100`

Output:
`ds4: prefill: 598.73 t/s, generation: 14.98 t/s`

When switching Spark to `--layers 22:42`: `ds4: prefill: 592.16 t/s, generation: 15.81 t/s`

This so far establishes:
1. Heterogenous distributed inference works across CUDA and Metal
2. Spreading prefill across two devices is faster even on 5GbE with 1-2ms ping, it doesn't need Thunderbolt
3. Generation is slow as expected

If this idea works out, I should be able to see ~600 t/s prefills (due to pipelining) and 30+ t/s generation (due to local generation).

**Comment by lobanov** [comment]
See #343 draft PR for early implementation.

Current status: usable for evaluation, but still largely research-focused work. Generation with local decode can work with ds4 and ds4-server' frontends. There is also work-in-progress implementation of ds4-evalfrontend for quality controls. Code compiles formetalandcuda-spark` targets.

Current focus of work: evaluation of quality of local decode after distributed prefill. Official logits gate passes, but I see divergent generation trajectories. I'm focusing on longer evaluations using modified ds4-eval. After that I will look at options to speed up the distributed handoff, as I sense lots of improvement opportunities.

How to run:
1. Connect two nodes with high-speed link. I use USB-C-to-5GbE adapter on Mac side and a short Cat8 cable plugged into DGX Spark, but I have not seen traffic more than 200MB/s, so 2.5GbE with Cat6 cable should be adequate, 1GbE will be too tight unless you switch to `--dist-activation-bits 16` (not tested).
2. Run controller on the node that will own early layers prefill, e.g. `./ds4 --role coordinator --layers 0:21 --listen 0.0.0.0 1234 --prompt-file speed-bench/promessi_sposi_1k.txt --ctx 200000 -n 1000` where `speed-bench/promessi_sposi_1k.txt` is the first 1000 lines of `speed-bench/promessi_sposi.txt` created by `head -n 1000 speed-bench/promessi_sposi.txt > speed-bench/promessi_sposi_1k.txt`, it's around 18k tokens. I use DGX Spark for this role. Only the owned layers (0-21 in this case) will be loaded into this instance.
3. Run controller on the node that will own later layers prefill and local generation, e.g. `./ds4 --role worker --layers 22:output --coordinator <coordinator_ip> 1234 --local-decode --ctx 200000`. I use MacBook M5 Max 128GB for that. Note that the worker with `--local-decode` must have enough memory to fit the entire model plus context.

On this reference setup I'm currently getting `ds4: prefill: 602.78 t/s, generation: 30.10 t/s`. Prefill speed is consistent with running fully distributed prefill acrss DGX Spark and M5 Max, and with fully local generation on M5 Max. Best of both worlds!

The implementation hands off KV cache for early layers to the node owning local decode, and catches of KV cache of early layers once the decode produces tokens, so repeated generation can resume correctly.

**Comment by lobanov** [comment]
PR #401 contains implementation of this feature.

**Comment by ngt1999** [comment]
Some data points on a simpler variant of this idea: instead of a live pipeline/handoff (#343 / #401), prefill on the Spark, copy the resulting disk KV checkpoint (`--kv-disk-dir`) to the Mac, and let the Mac's `ds4-server` pick it up as a normal disk-cache hit. No code changes were needed: the on-disk payload is already backend-neutral (Metal expands its F16 compressed cache to F32 when saving), so a CUDA-written `.kv` loads on Metal and vice versa.

**Setup**
- ds4 main `0aaea5a` (2026-09-20) on both machines, same GGUF (identical file size): `DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix-0731.gguf`
- ASUS Ascent GX10 (GB10, CUDA) and MacBook Pro 16" M5 Max 128 GB (Metal)
- Mac 5GbE (USB adapter) ↔ 10GbE switch ↔ GX10; checkpoints copied with `scp` (~500 MiB/s for files above ~450 MiB)
- Both servers: `--ctx 131072 --kv-disk-dir ~/ds4-kv --kv-disk-space-mb 32768 --kv-cache-cold-max-tokens 110000 --kv-cache-continued-interval-tokens 0 --kv-cache-reject-different-quant`, no MTP/DSpark
- Prompt: system = Promessi Sposi text sized to N tokens, user = short question. The cold checkpoint is stored at the user-turn anchor. The measured request uses a different question, so it can't hit live KV and must load from disk. `temperature 0`, thinking off.

**Results** (one run each; in all 16 cases the full checkpoint was loaded, `cached_tokens` = checkpoint tokens)

| input tokens | Mac cold prefill | Spark cold prefill | copy (size) | Mac TTFT, hit on Spark-made `.kv` | Mac time-to-ready: local vs Spark→Mac |
|---|---|---|---|---|---|
| 16.5k | 24.2 s | 18.6 s | 0.58 s (239 MiB) | 0.30 s | 24.2 s vs 19.5 s (1.24×) |
| 33.6k | 64.0 s | 37.5 s | 0.92 s (464 MiB) | 0.33 s | 64.0 s vs 38.7 s (1.65×) |
| 67.4k | 127.0 s | 76.4 s | 1.91 s (907 MiB) | 0.41 s | 127.0 s vs 78.7 s (1.61×) |
| 102.4k | 221.8 s | 117.1 s | 2.75 s (1367 MiB) | 0.50 s | 221.8 s vs 120.3 s (1.84×) |

**Observations**
- Spark prefill stays at ~870–900 t/s up to 100k, while the Mac drops from ~680 to ~460 t/s, so the gain grows with prompt length. Checkpoint size is ~13.8 KB/token.
- On the Mac, a hit on a Spark-made `.kv` has the same TTFT as a hit on a Mac-made one (0.50 s vs 0.50 s at 100k). Decode after the hit was also the same at 16k and 32k (39.7 and 32.4 t/s in both cases).
- At 65k and 100k, Mac decode after the cross hit was *faster* than after a local cold prefill (34.7 vs 29.7 t/s, and 31.9 vs 27.3 t/s). My guess is thermals: in the local case the Mac starts decoding right after 2–4 minutes of full-load prefill, while in the cross case it sat idle during the Spark prefill. Not verified yet.
- Caveat: hits were measured right after the file was written or copied, so the `.kv` was probably still in the page cache.
- I haven't evaluated output quality yet. Cross-backend numerics aren't bit-identical, as the ds4-eval run in #343 showed for the handoff path.

**Question:** is the disk KV format intended to stay portable across backends and machines? The header doesn't identify the weights (#805 / #882), so for now this relies on the user keeping the GGUF identical on both sides.

Script used (stdlib Python; drives two running `ds4-server`s, copies the checkpoint over ssh, records sizes and transfer times): <link>
